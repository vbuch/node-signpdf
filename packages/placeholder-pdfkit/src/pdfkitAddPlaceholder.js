import {
    DEFAULT_BYTE_RANGE_PLACEHOLDER,
    DEFAULT_SIGNATURE_LENGTH,
    SIG_FLAGS,
    SUBFILTER_ADOBE_PKCS7_DETACHED,
} from '@signpdf/utils';

/**
* @typedef {object} InputType
* @property {object} pdf PDFDocument
* @property {Buffer} pdfBuffer
* @property {string} reason
* @property {string} contactInfo
* @property {string} name
* @property {string} location
* @property {Date} [signingTime]
* @property {number} [signatureLength]
* @property {string} [byteRangePlaceholder]
* @property {string} [subFilter] One of SUBFILTER_* from \@signpdf/utils
* @property {number[]} [widgetRect] [x1, y1, x2, y2] widget rectangle
*/

/**
* @typedef {object} ReturnType
* @property {any} signature
* @property {any} form
* @property {any} widget
 */

/**
 * Adds the objects that are needed for Adobe.PPKLite to read the signature.
 * Also includes a placeholder for the actual signature.
 * Returns an Object with all the added PDFReferences.
 * @param {InputType}
 * @returns {ReturnType}
 */
export const pdfkitAddPlaceholder = ({
    pdf,
    reason,
    contactInfo,
    name,
    location,
    signingTime = undefined,
    signatureLength = DEFAULT_SIGNATURE_LENGTH,
    byteRangePlaceholder = DEFAULT_BYTE_RANGE_PLACEHOLDER,
    subFilter = SUBFILTER_ADOBE_PKCS7_DETACHED,
    widgetRect = [0, 0, 0, 0],
}) => {
    /* eslint-disable no-underscore-dangle,no-param-reassign */
    // Generate the signature placeholder
    const signature = pdf.ref({
        Type: 'Sig',
        Filter: 'Adobe.PPKLite',
        SubFilter: subFilter,
        ByteRange: [
            0,
            byteRangePlaceholder,
            byteRangePlaceholder,
            byteRangePlaceholder,
        ],
        Contents: Buffer.from(String.fromCharCode(0).repeat(signatureLength)),
        Reason: new String(reason), // eslint-disable-line no-new-wrappers
        M: signingTime ?? new Date(),
        ContactInfo: new String(contactInfo), // eslint-disable-line no-new-wrappers
        Name: new String(name), // eslint-disable-line no-new-wrappers
        Location: new String(location), // eslint-disable-line no-new-wrappers
    });

    if (!pdf._acroform) {
        pdf.initForm();
    }

    const form = pdf._root.data.AcroForm;
    const fieldId = form.data.Fields.length + 1;
    const signatureName = `Signature${fieldId}`;
    form.data = {
        Type: 'AcroForm',
        SigFlags: SIG_FLAGS.SIGNATURES_EXIST | SIG_FLAGS.APPEND_ONLY,
        Fields: form.data.Fields,
        DR: form.data.DR,
    };

    // Generate signature annotation widget
    const widget = pdf.ref({
        Type: 'Annot',
        Subtype: 'Widget',
        FT: 'Sig',
        Rect: widgetRect,
        V: signature,
        T: new String(signatureName), // eslint-disable-line no-new-wrappers
        P: pdf.page.dictionary,
    });

    pdf.page.annotations.push(widget);
    form.data.Fields.push(widget);
    signature.end();
    widget.end();

    return {
        signature,
        form,
        widget,
    };
    /* eslint-enable no-underscore-dangle,no-param-reassign */
};
