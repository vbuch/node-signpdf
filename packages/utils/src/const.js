export const DEFAULT_SIGNATURE_LENGTH = 8192;
export const DEFAULT_BYTE_RANGE_PLACEHOLDER = '**********';

export const SUBFILTER_ADOBE_PKCS7_DETACHED = 'adbe.pkcs7.detached';
export const SUBFILTER_ADOBE_PKCS7_SHA1 = 'adbe.pkcs7.sha1';
export const SUBFILTER_ADOBE_X509_SHA1 = 'adbe.x509.rsa.sha1';
export const SUBFILTER_ETSI_CADES_DETACHED = 'ETSI.CAdES.detached';

/**
 * Signature flags (bitmask) to be used under /SigFlags.
 *
 * {@link https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/pdfreference1.3.pdf}
 * See TABLE 7.42 and 7.43
 * @readonly
 * @enum {number}
 */
export const SIG_FLAGS = {
    /**
     * If set, the document contains at least one signature field.
     */
    SIGNATURES_EXIST: 1,
    /**
     * If set, the document contains signatures that may be invalidated
     * if the file is saved (written) in a way that alters its previous contents.
     */
    APPEND_ONLY: 2,
};

/**
 * Annotation flags (bitmask) to be used in /F under /Annot
 *
 * @readonly
 * @enum {number}
 * {@link https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/pdfreference1.3.pdf}
 * See TABLE 7.10
 */
export const ANNOTATION_FLAGS = {
    /**
     * If set, do not display the annotation if it does not belong to one of the
     * standard annotation types and no annotation handler is available.
     */
    INVISIBLE: 1,

    /**
     * If set, do not display or print the annotation or allow it to interact with the user,
     * regardless of its annotation type or whether an annotation handler is available.
     */
    HIDDEN: 2,

    /**
     * If set, print the annotation when the page is printed. If clear, never print the
     * annotation, regardless of whether it is displayed on the screen.
     */
    PRINT: 4,

    /**
     * If set, do not scale the annotation’s appearance to match the magnification of the page.
     */
    NO_ZOOM: 8,

    /**
     * If set, do not rotate the annotation’s appearance to match the rotation of the page.
     */
    NO_ROTATE: 16,

    /**
     * If set, do not display the annotation on the screen or allow it to interact with the user.
     */
    NO_VIEW: 32,

    /**
     * If set, do not allow the annotation to interact with the user.
     */
    READ_ONLY: 64,
};
