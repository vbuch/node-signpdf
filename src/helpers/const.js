export const DEFAULT_SIGNATURE_LENGTH = 8192;
export const DEFAULT_BYTE_RANGE_PLACEHOLDER = '**********';
export const SUBFILTER_ADOBE_PKCS7_DETACHED = 'adbe.pkcs7.detached';
export const SUBFILTER_ADOBE_PKCS7_SHA1 = 'adbe.pkcs7.sha1';
export const SUBFILTER_ADOBE_X509_SHA1 = 'adbe.x509.rsa.sha1';
export const SUBFILTER_ETSI_CADES_DETACHED = 'ETSI.CAdES.detached';
