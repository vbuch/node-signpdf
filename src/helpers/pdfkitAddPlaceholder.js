import {DEFAULT_BYTE_RANGE_PLACEHOLDER, DEFAULT_SIGNATURE_LENGTH, SUBFILTER_ADOBE_PKCS7_DETACHED} from './const';
// eslint-disable-next-line import/no-unresolved
import PDFKitReferenceMock from './pdfkitReferenceMock';
/**
 * Adds the objects that are needed for Adobe.PPKLite to read the signature.
 * Also includes a placeholder for the actual signature.
 * Returns an Object with all the added PDFReferences.
 * @param {PDFDocument} pdf
 * @param {string} reason
 * @returns {object}
 */
const pdfkitAddPlaceholder = ({
    pdf,
    pdfBuffer,
    reason,
    contactInfo = 'emailfromp1289@gmail.com',
    name = 'Name from p12',
    location = 'Location from p12',
    signatureLength = DEFAULT_SIGNATURE_LENGTH,
    byteRangePlaceholder = DEFAULT_BYTE_RANGE_PLACEHOLDER,
    subFilter = SUBFILTER_ADOBE_PKCS7_DETACHED,
}) => {
    /* eslint-disable no-underscore-dangle,no-param-reassign */
    // Generate the signature placeholder
    const signature = pdf.ref({
        Type: 'Sig',
        Filter: 'Adobe.PPKLite',
        SubFilter: subFilter,
        ByteRange: [
            0,
            byteRangePlaceholder,
            byteRangePlaceholder,
            byteRangePlaceholder,
        ],
        Contents: Buffer.from(String.fromCharCode(0).repeat(signatureLength)),
        Reason: new String(reason), // eslint-disable-line no-new-wrappers
        M: new Date(),
        ContactInfo: new String(contactInfo), // eslint-disable-line no-new-wrappers
        Name: new String(name), // eslint-disable-line no-new-wrappers
        Location: new String(location), // eslint-disable-line no-new-wrappers
    });

    // Check if pdf already contains acroform field
    const acroFormPosition = pdfBuffer.lastIndexOf('/Type /AcroForm');
    const isAcroFormExists = acroFormPosition !== -1;
    let fieldIds = [];
    let acroFormId;

    if (isAcroFormExists) {
        let acroFormStart = acroFormPosition;
        // 10 is the distance between "/Type /AcroForm" and AcroFrom ID
        const charsUntilIdEnd = 10;
        const acroFormIdEnd = acroFormPosition - charsUntilIdEnd;
        // Let's find AcroForm ID by trying to find the "\n" before the ID
        // 12 is a enough space to find the "\n"
        // (generally it's 2 or 3, but I'm giving a big space though)
        const maxAcroFormIdLength = 12;
        let foundAcroFormId = '';
        let index = charsUntilIdEnd + 1;
        for (index; index < charsUntilIdEnd + maxAcroFormIdLength; index += 1) {
            const acroFormIdString = pdfBuffer
                .slice(acroFormPosition - index, acroFormIdEnd).toString();

            if (acroFormIdString[0] === '\n') {
                break;
            }

            foundAcroFormId = acroFormIdString;
            acroFormStart = acroFormPosition - index;
        }

        const pdfSlice = pdfBuffer.slice(acroFormStart);
        const acroForm = pdfSlice.slice(0, pdfSlice.indexOf('endobj')).toString();
        acroFormId = parseInt(foundAcroFormId);

        const acroFormFields = acroForm.slice(acroForm.indexOf('/Fields [') + 9, acroForm.indexOf(']'));
        fieldIds = acroFormFields
            .split(' ')
            .filter((element, i) => i % 3 === 0)
            .map((fieldId) => new PDFKitReferenceMock(fieldId));
    }
    const signatureName = 'Signature';

    // Generate signature annotation widget
    const widget = pdf.ref({
        Type: 'Annot',
        Subtype: 'Widget',
        FT: 'Sig',
        Rect: [0, 0, 0, 0],
        V: signature,
        T: new String(signatureName + (fieldIds.length + 1)), // eslint-disable-line no-new-wrappers
        F: 4,
        P: pdf.page.dictionary, // eslint-disable-line no-underscore-dangle
    });
    pdf.page.dictionary.data.Annots = [widget];
    // Include the widget in a page
    let form;

    if (!isAcroFormExists) {
        // Create a form (with the widget) and link in the _root
        form = pdf.ref({
            Type: 'AcroForm',
            SigFlags: 3,
            Fields: [...fieldIds, widget],
        });
    } else {
        // Use existing acroform and extend the fields with newly created widgets
        form = pdf.ref({
            Type: 'AcroForm',
            SigFlags: 3,
            Fields: [...fieldIds, widget],
        }, acroFormId);
    }
    pdf._root.data.AcroForm = form;

    return {
        signature,
        form,
        widget,
    };
    /* eslint-enable no-underscore-dangle,no-param-reassign */
};

export default pdfkitAddPlaceholder;
