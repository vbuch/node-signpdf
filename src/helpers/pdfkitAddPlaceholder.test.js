import PDFDocument from 'pdfkit';
import pdfkitAddPlaceholder from './pdfkitAddPlaceholder';
import {SUBFILTER_ETSI_CADES_DETACHED} from './const';
import PDFObject from './pdfkit/pdfobject';

describe('pdfkitAddPlaceholder', () => {
    it('adds placeholder to PDFKit document', () => {
        const pdf = new PDFDocument({
            autoFirstPage: true,
            size: 'A4',
            layout: 'portrait',
            bufferPages: true,
        });
        pdf.info.CreationDate = '';

        const refs = pdfkitAddPlaceholder({pdf, pdfBuffer: Buffer.from([pdf])});
        expect(Object.keys(refs)).toMatchSnapshot();
        expect(pdf.page.dictionary.data.Annots).toHaveLength(1);
        expect(pdf.page.dictionary.data.Annots[0].data.Subtype).toEqual('Widget');
        expect(pdf.page.dictionary.data.Annots[0].data.V.data.ByteRange).toEqual([
            0,
            '**********',
            '**********',
            '**********',
        ]);
    });

    it('placeholder contains reason, contactInfo, name, location', () => {
        const pdf = new PDFDocument({
            autoFirstPage: true,
            size: 'A4',
            layout: 'portrait',
            bufferPages: true,
        });
        pdf.info.CreationDate = '';

        const refs = pdfkitAddPlaceholder({
            pdf,
            pdfBuffer: Buffer.from([pdf]),
            reason: 'test reason',
            contactInfo: 'testemail@test.com',
            name: 'test name',
            location: 'test Location',
        });
        expect(Object.keys(refs)).toMatchSnapshot();
        expect(pdf.page.dictionary.data.Annots).toHaveLength(1);
        expect(pdf.page.dictionary.data.Annots[0].data.Subtype).toEqual('Widget');
        const widgetData = pdf.page.dictionary.data.Annots[0].data.V.data;
        expect(PDFObject.convert(widgetData.Reason)).toEqual('(test reason)');
        expect(PDFObject.convert(widgetData.ContactInfo)).toEqual('(testemail@test.com)');
        expect(PDFObject.convert(widgetData.Name)).toEqual('(test name)');
        expect(PDFObject.convert(widgetData.Location)).toEqual('(test Location)');
    });

    it('placeholder contains default values for contactInfo, name, location', () => {
        const pdf = new PDFDocument({
            autoFirstPage: true,
            size: 'A4',
            layout: 'portrait',
            bufferPages: true,
        });
        pdf.info.CreationDate = '';

        const refs = pdfkitAddPlaceholder({
            pdf,
            pdfBuffer: Buffer.from([pdf]),
            reason: 'test reason',
        });
        expect(Object.keys(refs)).toMatchSnapshot();
        expect(pdf.page.dictionary.data.Annots).toHaveLength(1);
        expect(pdf.page.dictionary.data.Annots[0].data.Subtype).toEqual('Widget');
        const widgetData = pdf.page.dictionary.data.Annots[0].data.V.data;
        expect(PDFObject.convert(widgetData.Reason)).toEqual('(test reason)');
        expect(PDFObject.convert(widgetData.ContactInfo)).toEqual('(emailfromp1289@gmail.com)');
        expect(PDFObject.convert(widgetData.Name)).toEqual('(Name from p12)');
        expect(PDFObject.convert(widgetData.Location)).toEqual('(Location from p12)');
        expect(widgetData.SubFilter).toEqual('adbe.pkcs7.detached');
    });

    it('allows defining signature SubFilter', () => {
        const pdf = new PDFDocument({
            autoFirstPage: true,
            size: 'A4',
            layout: 'portrait',
            bufferPages: true,
        });
        pdf.info.CreationDate = '';

        const refs = pdfkitAddPlaceholder({
            pdf,
            pdfBuffer: Buffer.from([pdf]),
            reason: 'test reason',
            subFilter: SUBFILTER_ETSI_CADES_DETACHED,
        });
        expect(Object.keys(refs)).toMatchSnapshot();
        expect(pdf.page.dictionary.data.Annots).toHaveLength(1);
        const widget = pdf.page.dictionary.data.Annots[0];
        expect(widget.data.Subtype).toEqual('Widget');
        const widgetData = widget.data.V.data;
        expect(PDFObject.convert(widgetData.Reason)).toEqual('(test reason)');
        expect(widgetData.SubFilter).toEqual('ETSI.CAdES.detached');
    });
});
