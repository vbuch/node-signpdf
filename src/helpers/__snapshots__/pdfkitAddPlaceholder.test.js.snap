// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pdfkitAddPlaceholder adds placeholder to PDFKit document 1`] = `
Array [
  "signature",
  "form",
  "widget",
]
`;

exports[`pdfkitAddPlaceholder allows defining signature SubFilter 1`] = `
Array [
  "signature",
  "form",
  "widget",
]
`;

exports[`pdfkitAddPlaceholder placeholder contains default values for contactInfo, name, location 1`] = `
Array [
  "signature",
  "form",
  "widget",
]
`;

exports[`pdfkitAddPlaceholder placeholder contains reason, contactInfo, name, location 1`] = `
Array [
  "signature",
  "form",
  "widget",
]
`;
